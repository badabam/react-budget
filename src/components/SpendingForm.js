import React, { Component, createRef } from 'react'
import styled from 'react-emotion'
import StyledButton from './StyledButton'
import StyledInput from './StyledInput'

const StyledForm = styled('form')`
  grid-row: 3;
  padding: 0 10px 20px;
  display: grid;
  grid-gap: 10px;
`

export default class SpendingForm extends Component {
  constructor(props) {
    super(props)
    this.input = createRef()
  }

  handleSubmit = e => {
    e.preventDefault()
    this.props.onSubmit()
    this.input.current && this.input.current.focus()
  }

  render() {
    const { textValue, amountValue, updateFormInput } = this.props

    return (
      <StyledForm onSubmit={this.handleSubmit}>
        <StyledInput
          placeholder="Name"
          onChange={e =>
            updateFormInput({
              name: 'textValue',
              value: e.target.value,
            })
          }
          innerRef={this.input}
          value={textValue}
        />
        <StyledInput
          placeholder="Amount"
          type="number"
          onChange={e =>
            updateFormInput({
              name: 'amountValue',
              value: e.target.value,
            })
          }
          value={amountValue}
        />
        <StyledButton
          disabled={textValue.trim() === '' || amountValue.trim() === ''}
        >
          Spend
        </StyledButton>
      </StyledForm>
    )
  }
}
