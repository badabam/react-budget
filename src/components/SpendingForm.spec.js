import React from 'react'
import SpendingForm from './SpendingForm'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

describe('SpendingForm', () => {
  let wrapper
  let submitFunc

  beforeEach(() => {
    submitFunc = jest.fn()
    wrapper = shallow(
      <SpendingForm textValue="foo" amountValue={3} onSubmit={submitFunc} />
    )
  })

  describe('visuals', () => {
    beforeEach(() => {
      wrapper = mount(
        <SpendingForm textValue="foo" amountValue={3} onSubmit={submitFunc} />
      )
    })

    it('has two inputs', () => {
      expect(wrapper.find('input').length).toBe(2)
    })

    it('has one button', () => {
      expect(wrapper.find('button').length).toBe(1)
    })
  })

  describe('prop changes', () => {
    it('updates its textValue', () => {
      expect(toJson(wrapper)).toMatchSnapshot()
      wrapper.setProps({ textValue: 'bar' })
      expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('updates its amountValue', () => {
      expect(toJson(wrapper)).toMatchSnapshot()
      wrapper.setProps({ inputValue: 1000 })
      expect(toJson(wrapper)).toMatchSnapshot()
    })
  })
})
