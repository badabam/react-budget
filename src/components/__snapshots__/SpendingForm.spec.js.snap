// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpendingForm prop changes updates its amountValue 1`] = `
<Styled(form)
  onSubmit={[Function]}
>
  <Styled(input)
    innerRef={
      Object {
        "current": null,
      }
    }
    onChange={[Function]}
    placeholder="Name"
    value="foo"
  />
  <Styled(Component)
    onChange={[Function]}
    onInit={[Function]}
    options={
      Object {
        "numeral": true,
        "prefix": "€ ",
        "rawValueTrimPrefix": true,
      }
    }
    placeholder="Amount"
    value={3}
  />
  <Styled(button)
    disabled={false}
  >
    Spend
  </Styled(button)>
</Styled(form)>
`;

exports[`SpendingForm prop changes updates its amountValue 2`] = `
<Styled(form)
  onSubmit={[Function]}
>
  <Styled(input)
    innerRef={
      Object {
        "current": null,
      }
    }
    onChange={[Function]}
    placeholder="Name"
    value="foo"
  />
  <Styled(Component)
    onChange={[Function]}
    onInit={[Function]}
    options={
      Object {
        "numeral": true,
        "prefix": "€ ",
        "rawValueTrimPrefix": true,
      }
    }
    placeholder="Amount"
    value={3}
  />
  <Styled(button)
    disabled={false}
  >
    Spend
  </Styled(button)>
</Styled(form)>
`;

exports[`SpendingForm prop changes updates its textValue 1`] = `
<Styled(form)
  onSubmit={[Function]}
>
  <Styled(input)
    innerRef={
      Object {
        "current": null,
      }
    }
    onChange={[Function]}
    placeholder="Name"
    value="foo"
  />
  <Styled(Component)
    onChange={[Function]}
    onInit={[Function]}
    options={
      Object {
        "numeral": true,
        "prefix": "€ ",
        "rawValueTrimPrefix": true,
      }
    }
    placeholder="Amount"
    value={3}
  />
  <Styled(button)
    disabled={false}
  >
    Spend
  </Styled(button)>
</Styled(form)>
`;

exports[`SpendingForm prop changes updates its textValue 2`] = `
<Styled(form)
  onSubmit={[Function]}
>
  <Styled(input)
    innerRef={
      Object {
        "current": null,
      }
    }
    onChange={[Function]}
    placeholder="Name"
    value="bar"
  />
  <Styled(Component)
    onChange={[Function]}
    onInit={[Function]}
    options={
      Object {
        "numeral": true,
        "prefix": "€ ",
        "rawValueTrimPrefix": true,
      }
    }
    placeholder="Amount"
    value={3}
  />
  <Styled(button)
    disabled={false}
  >
    Spend
  </Styled(button)>
</Styled(form)>
`;
